src/renderer.ts

```diff
@@ -156,6 +158,632 @@ export class DefaultCellRenderer implements ICellRendererFactory {
+
+      var index = (percentile / 100) * data.length;
+      var result;
+      if (Math.floor(index) === index) {
+        result = (data[(index - 1)] + data[index]) / 2;
+      } else {
+        result = data[Math.floor(index)];
+      }
+      return result;
+    }
+
+    function numSort(a, b) {
+      return a - b;
+    }
+
+    const padding = context.option('rowPadding', 1);
+    const min = (<any> col.desc).domain[0];
+    const max = (<any> col.desc).domain[1];
+    var scale = d3.scale.linear().domain([min, max]).range([0, col.getWidth()]); // Constraint the window width
+    return (ctx: CanvasRenderingContext2D, d: IDataRow, i: number) => {
+      // Rectangle
+
+      var data = (col.getValue(d.v, i));
+
+      data.sort(numSort);
+      var mindata = Math.min.apply(Math, data);
+      var maxdata = Math.max.apply(Math, data);
+      var q1 = (getPercentile(data, 25));
+      var med = (getPercentile(data, 50));
+      var q3 = (getPercentile(data, 75));
+      ctx.fillStyle = '#e0e0e0';
+      ctx.strokeStyle = 'black';
+      ctx.beginPath();
+      ctx.rect(scale(q1), padding, (scale(q3) - scale(q1)), context.rowHeight(i));
+      ctx.fill();
+      ctx.stroke();
+
+
+      //Line
+      var left = scale(mindata), right = scale(maxdata), center = scale(med);
+      var top = context.option('rowPadding', 1);
+      var bottom = Math.max(context.rowHeight(i) - top, 0);
+      var middle = (bottom - top) / 2;
+      ctx.strokeStyle = 'black';
+      ctx.fillStyle = '#e0e0e0';
+      ctx.beginPath();
+      ctx.moveTo(left, middle);
+      ctx.lineTo(scale(q1), middle);
+      ctx.moveTo(left, top);
+      ctx.lineTo(left, bottom);
+      ctx.moveTo(center, top);
+      ctx.lineTo(center, bottom);
+      ctx.moveTo((scale(q1) + scale(q3) - scale(q1)), middle);
+      ctx.lineTo(right, middle);
+      ctx.moveTo(right, top);
+      ctx.lineTo(right, bottom);
+      ctx.stroke();
+      ctx.fill();
+
+    };
+  }
+
+
+}
+
+class UpsetCellRenderer extends DefaultCellRenderer {
+
+  createSVG(col: Column, context: IDOMRenderContext): ISVGCellRenderer {
+
+
+    const bins = (<any> col.desc).datalength;
+    const windowsize = col.getWidth() / bins;
+
+
+    return {
+      template: `<g class="upsetcell"></g>`,
+      update: (n: SVGGElement, d: IDataRow, i: number) => {
+        const g = d3.select(n);
+        const circle = g.selectAll('circle').data(<any>col.getValue(d.v, i));
+        circle.enter().append('circle');
+        circle
+          .attr({
+            cy: (d: any, i) => (context.rowHeight(i) / 2),
+            cx: (d: any, i) => (i * windowsize) + (windowsize / 2),
+            r: (windowsize / 4),
+            class: 'upsetcircle',
+            opacity: (d) =>(d === 1) ? 1 : 0.1
+          });
+        circle.exit().remove();
+        var catindexes = [];
+        const path = g.selectAll('path').data(<any>[col.getValue(d.v, i)]);
+        var countcategory = col.getValue(d.v, i).filter((x) => x === 1).length;
```

Comment:
assuming magic constant `1` to be something special. constant assumptions should be constant variables and in this case moved the model class.